### PR TITLE
[Example] Add examples for rgcn edge classification

### PR DESCRIPTION
**Pull request by hellomirro**
## Description
Relational GCN for edge classification
Code follows the dgl tutorial:
https://docs.dgl.ai/guide/minibatch-edge.html#guide-minibatch-edge-classification-sampler
https://github.com/dmlc/dgl/blob/master/examples/pytorch/rgcn/entity_classify_mp.py

Data is generated by fake_data_generator.py
Graph: 2 types of nodes (seller and product). A seller-product pair is an offer listing, ie. edge. Task is to predict
whether an offer listing (edge) is defective.
Model: use RGCN propagation to get node embeddings, and then concatenate node embeddings with edge embeddings to make a
binary classification. The edge embeddings are generated from a neural network. All parameters are optimized as a whole
by the cross entropy objective function.
This code can handle:
1. Features on different node type have different dimensions.
2. Mini-batch training
3. RGCN homogeneous implementation.
4. Model performance is evaluated by accuracy and roc auc.

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage
- [ ] Code is well-documented
- [ ] To the my best knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

File: examples/pytorch/rgcn/edge_classify.py [hunk 1]
```diff
@@ -0,0 +1,490 @@
...
+import torch.nn as nn
+import torch.nn.functional as F
+import dgl
+from dgl.nn.pytorch import RelGraphConv
+from sklearn.metrics import roc_auc_score
+
+from model import RelGraphEmbedLayer
+from fake_data_generator import load_generated_data
+
+class RGCN(nn.Module):
```
Comment (classicsong) [inline examples/pytorch/rgcn/edge_classify.py:38]: Can we re-use the module from the node classification example?

File: examples/pytorch/rgcn/edge_classify.py [hunk 2]
```diff
@@ -0,0 +1,490 @@
...
+class NeighborSampler:
+    """Neighbor sampler
+    Parameters
+    ----------
+    fanouts : list of int
+        Fanout of each hop starting from the seed nodes. If a fanout is None,
+        sample full neighbors.
+    """
+    def __init__(self, fanouts):
+
```
Comment (classicsong) [inline examples/pytorch/rgcn/edge_classify.py:128]: empty line here.

File: examples/pytorch/rgcn/edge_classify.py [hunk 3]
```diff
@@ -0,0 +1,490 @@
...
+class ScorePredictor(nn.Module):
+    def __init__(self, num_classes, in_features, edge_features):
+        super().__init__()
+        self.edge_features = edge_features
+
+        edge_feature_dim = edge_features.shape[1]
+        h_dim = int(edge_feature_dim / 2)
+        print("edge_feature_dim: ", edge_feature_dim)
+        print("h_dim: ", h_dim)
+        dropout = 0.2
```
Comment (classicsong) [inline examples/pytorch/rgcn/edge_classify.py:173]: Can we make dropout configuration?

File: examples/pytorch/rgcn/edge_classify.py [hunk 4]
```diff
@@ -0,0 +1,490 @@
...
+            th.nn.Linear(edge_feature_dim, h_dim),
+            th.nn.ReLU(),
+            th.nn.Dropout(dropout)
+        )
+
+        self.linear = nn.Linear(2 * in_features + h_dim, num_classes)
+
+    def apply_edges(self, edges):
+        offer_id = edges.data['type_id'].tolist()
+        offer_features = self.edge_features[offer_id]
```
Comment (classicsong) [inline examples/pytorch/rgcn/edge_classify.py:186]: Can we preprocess the edge_feature into torch tensor?
BTW, is this a common case to concatenate the edge feature? (Can we make this an optional choice?)

File: examples/pytorch/rgcn/edge_classify.py [hunk 5]
```diff
@@ -0,0 +1,490 @@
...
+def main(args):
+    # load graph data
+    hg, node_feats, edge_feats, label_edge = load_generated_data()
+
+    num_of_ntype = len(hg.ntypes)
+    num_rels = len(hg.etypes)
+
+    g = dgl.to_homogeneous(hg, edata=["label"])
+
+    etype_id = hg.get_etype_id(label_edge)
```
Comment (classicsong) [inline examples/pytorch/rgcn/edge_classify.py:421]: Can you add some comments here?

File: examples/pytorch/rgcn/fake_data_generator.py [hunk 6]
```diff
@@ -0,0 +1,201 @@
+import pandas as pd
+import numpy as np
+import torch as th
+import dgl
+
+class CreateFakeData:
```
Comment (classicsong) [inline examples/pytorch/rgcn/fake_data_generator.py:6]: Can we briefly describe the structure of the faked data?

**Review by classicsong — COMMENTED** [review]
We may need to find an edge classification dataset for this example. @zheng-da @BarclayII Any ideas?

How about this: https://www.kaggle.com/snap/amazon-fine-food-reviews

**Comment by BarclayII** [comment]
For edge classification with 2 types of nodes I think MovieLens is already a good example for demonstration purpose, though there are no papers verifying the effectiveness of RGCN on rating prediction in particular (GCMC can also be considered as a kind of RGCN by the way).

**Comment by hellomirro** [comment]
> We may need to find an edge classification dataset for this example. @zheng-da @BarclayII Any ideas?
> 
> How about this: https://www.kaggle.com/snap/amazon-fine-food-reviews

Thanks for reviewing my code. I am happy to apply this code to a real data set. Let me know which data set would be a good example, and how to define the nodes, edges, and prediction task.

File: examples/pytorch/rgcn/edge_classify.py (hunk 1 above)
Comment (hellomirro) [reply examples/pytorch/rgcn/edge_classify.py]: Done. Compared to your RGCN module in node classification example, I have a new arg "input feat_dim". Its dim is different from h_dim. In your example, your input feat dim == h_him.

File: examples/pytorch/rgcn/edge_classify.py (hunk 2 above)
Comment (hellomirro) [reply examples/pytorch/rgcn/edge_classify.py]: Done

File: examples/pytorch/rgcn/edge_classify.py (hunk 3 above)
Comment (hellomirro) [reply examples/pytorch/rgcn/edge_classify.py]: Done

File: examples/pytorch/rgcn/edge_classify.py (hunk 4 above)
Comment (hellomirro) [reply examples/pytorch/rgcn/edge_classify.py]: Done. Added an option if the edge_feature is None, then only use node features to make predictions.

File: examples/pytorch/rgcn/edge_classify.py (hunk 5 above)
Comment (hellomirro) [reply examples/pytorch/rgcn/edge_classify.py:418]: Done

File: examples/pytorch/rgcn/fake_data_generator.py (hunk 6 above)
Comment (hellomirro) [reply examples/pytorch/rgcn/fake_data_generator.py:54]: Done

**Comment by zheng-da** [comment]
I also think MovieLens is a good dataset for this.
